HerPortal.ManagementShell/AdminAction.cs

```diff
@@ -32,161 +29,90 @@ public AdminAction(IDatabaseOperation dbOperation, IOutputProvider outputProvide
-            return;
-        }
+            throw new CommandException(
+                $"Could not find LAs attached to {user.EmailAddress} for the following codes: {string.Join(", ", missingCodes)}. Please check your inputs and try again.");
 
         dbOperation.RemoveLasFromUser(user, lasToRemove);
     }
 
-    public void CreateOrUpdateUser(string? userEmailAddress, string[] custodianCodes)
+    public void AddConsortia(User user, IReadOnlyCollection<string> consortiumCodes)
+    {
+        var consortiaToAdd = dbOperation.GetConsortia(consortiumCodes);
+
+        var ownedCustodianCodesInConsortia = GetOwnedCustodianCodesInConsortia(user, consortiumCodes);
+        var lasToRemove = dbOperation.GetLas(ownedCustodianCodesInConsortia);
+
+        if (lasToRemove.Count > 0)
+            dbOperation.AddConsortiaAndRemoveLasFromUser(user, consortiaToAdd, lasToRemove);
+        else
+            dbOperation.AddConsortiaToUser(user, consortiaToAdd);
+    }
+
+    public void RemoveConsortia(User user, IReadOnlyCollection<string> consortiumCodes)
     {
-        if (userEmailAddress == null)
-        {
-            outputProvider.Output("Please specify user E-mail address to create or update");
-            return;
-        }
-
-        var user = GetUser(userEmailAddress);
-        var userStatus = GetUserStatus(user);
-        DisplayUserStatus(userStatus);
-        outputProvider.Output("");
-
-        outputProvider.Output("!!! ATTENTION! READ CAREFULLY OR RISK A DATA BREACH !!!");
-        outputProvider.Output("");
-        outputProvider.Output("You are about to grant a user permission to read PERSONALLY IDENTIFIABLE INFORMATION submitted to LAs.");
-        outputProvider.Output("Take a moment to double check the following list and only continue if you are certain this user should have access to these LAs.");
-        outputProvider.Output("NB: in particular, you should only do this for LAs that have signed their DSA contracts!");
-        outputProvider.Output("");
-
-        var confirmation = ConfirmCustodianCodes(userEmailAddress, custodianCodes);
-
-        if (confirmation)
-        {
-            if (userStatus.Equals(UserStatus.Active))
-            {
-                AddLas(user, custodianCodes);
-            }
-            else if (userStatus.Equals(UserStatus.New))
-            {
-                TryCreateUser(userEmailAddress, custodianCodes);
-            }
-        }
+        var consortiaToRemove = user.Consortia.Where(consortium => consortiumCodes.Contains(consortium.ConsortiumCode))
+            .ToList();
+        var missingCodes = consortiumCodes
+            .Where(code => !consortiaToRemove.Any(consortium => consortium.ConsortiumCode.Equals(code))).ToList();
+        if (missingCodes.Count > 0)
+            throw new CommandException(
+                $"Could not find Consortia attached to {user.EmailAddress} for the following codes: {string.Join(", ", missingCodes)}. Please check your inputs and try again.");
```

Comment:
Was going to say this last time but got myself tangled up about which class was doing what. I think setting the message here is getting a bit close to having this class take responsibility for writing to the console, which is what you took away from it.

Would be a cleaner boundary if it was completely up to the `CommandHandler` to write the message. You can do that by adding a property to the exception for a list of the missing codes and using that upstairs. It's probably a good idea to give `CommandException` a more specific name along with that too.

Reply:
have done some reshuffling now, most of the error message construction has now been moved upwards, with some more description at the top level

thought it still appropriate for the exception to contain a message parameter with the context of the exception. I think this should be the smallest amount of text the exception need expose, since the top level call can have a couple possible exceptions happen